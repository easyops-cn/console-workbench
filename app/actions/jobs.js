import { spawn } from 'child_process';
import { max } from 'lodash';
import { sync } from 'shell-env';
import defaultShell from 'default-shell';
import { addTask, findTaskByJobId, removeTaskByJobId } from '../utils/tasks';
import storage from '../storage';

export const ADD_JOB = 'ADD_JOB';
export const UPDATE_JOB = 'UPDATE_JOB';
export const REMOVE_JOB = 'REMOVE_JOB';
export const REPLACE_JOBS = 'REPLACE_JOBS';
export const START_JOB = 'START_JOB';
export const STARTED_JOB = 'STARTED_JOB';
export const STOP_JOB = 'STOP_JOB';
export const JOB_OUTPUT = 'JOB_OUTPUT';
export const JOB_CLOSE = 'JOB_CLOSE';
export const JOB_ERROR = 'JOB_ERROR';
export const JOB_EXIT = 'JOB_EXIT';
export const CLEAR_JOB_OUTPUT = 'CLEAR_JOB_OUTPUT';
export const ACTIVATE_JOB = 'ACTIVATE_JOB';

export const addJob = ({ name, cmd, cwd }) => dispatch => {
  const jobs = storage.get('jobs', []);
  const nextId = jobs.length === 0 ? 1 : max(jobs.map(item => item.id)) + 1;
  const newJob = {
    id: nextId,
    name,
    cmd,
    cwd
  };
  storage.set('jobs', [...jobs, newJob]);

  dispatch({
    type: ADD_JOB,
    job: newJob
  });
};

export const updateJob = ({ id, name, cmd, cwd }) => dispatch => {
  const job = { id, name, cmd, cwd };
  const jobs = storage.get('jobs');
  const originalJob = jobs.find(item => item.id === job.id);
  Object.assign(originalJob, job);
  storage.set('jobs', jobs);

  dispatch({
    type: UPDATE_JOB,
    job
  });
};

export const replaceJobs = jobs => dispatch => {
  storage.set('jobs', jobs);

  dispatch({
    type: REPLACE_JOBS,
    jobs
  });
};

export const removeJob = job => dispatch => {
  const jobs = storage.get('jobs');
  storage.set('jobs', jobs.filter(item => item.id !== job.id));

  dispatch({
    type: REMOVE_JOB,
    job
  });
};

const jobOutput = (job, output) => ({
  type: JOB_OUTPUT,
  job,
  output
});

const jobClose = (job, code, signal) => ({
  type: JOB_CLOSE,
  job,
  code,
  signal
});

const jobError = (job, error) => ({
  type: JOB_ERROR,
  job,
  error
});

const jobExit = (job, code, signal) => ({
  type: JOB_EXIT,
  job,
  code,
  signal
});

export const startJob = job => dispatch => {
  if (job.starting || job.running) {
    return;
  }

  dispatch({
    type: START_JOB,
    job
  });

  const task = spawn(job.cmd, [], {
    detached: true,
    cwd: job.cwd,
    env: sync(defaultShell),
    shell: defaultShell
  });

  addTask(job.id, task);

  dispatch({
    type: STARTED_JOB,
    job
  });

  task.stdout.on('data', data => {
    dispatch(jobOutput(job, data.toString()));
  });
  task.stderr.on('data', data => {
    dispatch(jobOutput(job, data.toString()));
  });
  task.on('close', (code, signal) => {
    removeTaskByJobId(job.id);
    dispatch(jobClose(job, code, signal));
  });
  task.on('error', error => {
    removeTaskByJobId(job.id);
    dispatch(jobError(job, error));
  });
  task.on('exit', (code, signal) => {
    removeTaskByJobId(job.id);
    dispatch(jobExit(job, code, signal));
  });
};

export const stopJob = job => dispatch => {
  if (!job.running || job.stopping) {
    return;
  }

  const task = findTaskByJobId(job.id);
  if (task === undefined) {
    return;
  }

  process.kill(-task.pid);

  return dispatch({
    type: STOP_JOB,
    job
  });
};

export const clearJobOutput = job => ({
  type: CLEAR_JOB_OUTPUT,
  job
});

export const activateJob = job => ({
  type: ACTIVATE_JOB,
  job
});
