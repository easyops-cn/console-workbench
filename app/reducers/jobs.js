// @flow
import { combineReducers } from 'redux';
import { without } from 'lodash';
import type { Action } from './types';
import {
  ADD_JOB,
  REMOVE_JOB,
  START_JOB,
  STARTED_JOB,
  JOB_OUTPUT,
  STOP_JOB,
  JOB_CLOSE,
  JOB_ERROR,
  JOB_EXIT,
  CLEAR_JOB_OUTPUT,
  ACTIVATE_JOB,
  UPDATE_JOB,
  REPLACE_JOBS
} from '../actions/jobs';
import printable from '../utils/printable';

const ids = (state = [], action: Action) => {
  switch (action.type) {
    case ADD_JOB:
      return [...state, action.job.id];
    case REMOVE_JOB:
      return without(state, action.job.id);
    case REPLACE_JOBS:
      return action.jobs.map(job => job.id);
    default:
      return state;
  }
};

const entities = (state = {}, action: Action) => {
  switch (action.type) {
    case ADD_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...action.job,
          starting: false,
          running: false,
          stopping: false
        }
      };
    case UPDATE_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          name: action.job.name,
          cmd: action.job.cmd,
          cwd: action.job.cwd,
          subPackageDir: action.job.subPackageDir
        }
      };
    case REMOVE_JOB:
      return {
        ...state,
        [action.job.id]: undefined
      };
    case REPLACE_JOBS:
      return action.jobs.reduce((acc, job) => {
        acc[job.id] = {
          ...job,
          starting: false,
          running: false,
          stopping: false
        };
        return acc;
      }, {});
    case START_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          starting: true,
          error: undefined
        }
      };
    case STARTED_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          starting: false,
          running: true
        }
      };
    case JOB_CLOSE:
    case JOB_EXIT:
      if (!state[action.job.id].running) {
        return state;
      }
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          running: false,
          stopping: false
        }
      };
    case JOB_ERROR:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          running: false,
          error: action.error
        }
      };
    case STOP_JOB:
      return {
        ...state,
        [action.job.id]: {
          ...state[action.job.id],
          stopping: true
        }
      };
    default:
      return state;
  }
};

const outputs = (state = {}, action) => {
  switch (action.type) {
    case ADD_JOB:
    case REMOVE_JOB:
    case CLEAR_JOB_OUTPUT:
      return {
        ...state,
        [action.job.id]: undefined
      };
    case REPLACE_JOBS:
      return action.jobs.reduce((acc, job) => {
        acc[job.id] = undefined;
        return acc;
      }, {});
    case START_JOB:
      return {
        ...state,
        [action.job.id]: printable(
          undefined,
          `$ cd ${action.job.cwd}\n$ ${action.job.cmd}\n`
        )
      };
    case JOB_OUTPUT:
      return {
        ...state,
        [action.job.id]: printable(state[action.job.id], action.output)
      };
    case JOB_CLOSE:
      return {
        ...state,
        [action.job.id]: printable(
          state[action.job.id],
          `closed with code: ${action.code}, signal: ${action.signal}\n`
        )
      };
    case JOB_EXIT:
      return {
        ...state,
        [action.job.id]: printable(
          state[action.job.id],
          `exited with code: ${action.code}, signal: ${action.signal}\n`
        )
      };
    case STOP_JOB:
      return {
        ...state,
        [action.job.id]: printable(state[action.job.id], '^C\n')
      };
    default:
      return state;
  }
};

const active = (state = null, action) => {
  switch (action.type) {
    case ACTIVATE_JOB:
      return state === action.job.id ? null : action.job.id;
    default:
      return state;
  }
};

const jobs = combineReducers({
  ids,
  entities,
  outputs,
  active
});

export default jobs;
