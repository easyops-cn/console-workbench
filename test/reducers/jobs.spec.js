import jobs from '../../app/reducers/jobs';
import {
  ADD_JOB,
  UPDATE_JOB,
  REMOVE_JOB,
  REPLACE_JOBS,
  START_JOB,
  STARTED_JOB,
  JOB_OUTPUT,
  JOB_CLOSE,
  JOB_EXIT,
  JOB_ERROR,
  STOP_JOB,
  CLEAR_JOB_OUTPUT,
  ACTIVATE_JOB
} from '../../app/actions/jobs';

describe('reducers', () => {
  describe('jobs', () => {
    it('should handle initial state', () => {
      expect(jobs(undefined, {})).toMatchSnapshot();
    });

    it('should handle ADD_JOB', () => {
      expect(
        jobs(undefined, {
          type: ADD_JOB,
          job: {
            id: 1,
            name: 'test job for ADD_JOB',
            cmd: 'ls -l',
            cwd: '/tmp'
          }
        })
      ).toMatchSnapshot();
    });

    it('should handle UPDATE_JOB', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for UPDATE_JOB',
                cmd: 'ls -l',
                cwd: '/tmp'
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: UPDATE_JOB,
            job: {
              id: 1,
              name: 'updated test job for UPDATE_JOB',
              cmd: 'ls -la',
              cwd: '/home'
            }
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle REMOVE_JOB', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for REMOVE_JOB',
                cmd: 'ls -l',
                cwd: '/tmp'
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: REMOVE_JOB,
            job: {
              id: 1,
              name: 'test job for REMOVE_JOB',
              cmd: 'ls -l',
              cwd: '/tmp'
            }
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle REPLACE_JOBS', () => {
      expect(
        jobs(undefined, {
          type: REPLACE_JOBS,
          jobs: [
            {
              id: 1,
              name: 'test job for REPLACE_JOBS',
              cmd: 'ls -l',
              cwd: '/tmp'
            }
          ]
        })
      ).toMatchSnapshot();
    });

    it('should handle START_JOB', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for START_JOB',
                cmd: 'ls -l',
                cwd: '/tmp'
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: START_JOB,
            job: {
              id: 1,
              name: 'test job for START_JOB',
              cmd: 'ls -l',
              cwd: '/tmp'
            }
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle STARTED_JOB', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for STARTED_JOB',
                cmd: 'ls -l',
                cwd: '/tmp',
                starting: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: STARTED_JOB,
            job: {
              id: 1,
              name: 'test job for STARTED_JOB',
              cmd: 'ls -l',
              cwd: '/tmp'
            }
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle JOB_OUTPUT', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for JOB_OUTPUT',
                cmd: 'ls -l',
                cwd: '/tmp',
                running: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: JOB_OUTPUT,
            job: {
              id: 1,
              name: 'test job for JOB_OUTPUT',
              cmd: 'ls -l',
              cwd: '/tmp'
            },
            output: '\n10% run\b\b\b\b\b\b\b100% done.\n'
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle JOB_CLOSE', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for JOB_CLOSE',
                cmd: 'ls -l',
                cwd: '/tmp',
                running: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: JOB_CLOSE,
            job: {
              id: 1,
              name: 'test job for JOB_CLOSE',
              cmd: 'ls -l',
              cwd: '/tmp'
            },
            code: 15,
            signal: 'SIGTERM'
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle JOB_EXIT', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for JOB_EXIT',
                cmd: 'ls -l',
                cwd: '/tmp',
                running: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: JOB_EXIT,
            job: {
              id: 1,
              name: 'test job for JOB_EXIT',
              cmd: 'ls -l',
              cwd: '/tmp'
            },
            code: 1,
            signal: 'SIGHUP'
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle JOB_ERROR', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for JOB_ERROR',
                cmd: 'ls -l',
                cwd: '/tmp',
                running: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: JOB_ERROR,
            job: {
              id: 1,
              name: 'test job for JOB_ERROR',
              cmd: 'ls -l',
              cwd: '/tmp'
            },
            error: new Error('error on purpose')
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle STOP_JOB', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for STOP_JOB',
                cmd: 'ls -l',
                cwd: '/tmp',
                running: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: STOP_JOB,
            job: {
              id: 1,
              name: 'test job for STOP_JOB',
              cmd: 'ls -l',
              cwd: '/tmp'
            }
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle CLEAR_JOB_OUTPUT', () => {
      expect(
        jobs(
          {
            ids: [1],
            entities: {
              1: {
                id: 1,
                name: 'test job for CLEAR_JOB_OUTPUT',
                cmd: 'ls -l',
                cwd: '/tmp',
                running: true
              }
            },
            outputs: {
              1: {
                returned: '$ ls -l',
                buffer: ['drwx'],
                cursor: 4
              }
            }
          },
          {
            type: CLEAR_JOB_OUTPUT,
            job: {
              id: 1,
              name: 'test job for CLEAR_JOB_OUTPUT',
              cmd: 'ls -l',
              cwd: '/tmp'
            }
          }
        )
      ).toMatchSnapshot();
    });

    it('should handle ACTIVATE_JOB', () => {
      expect(
        jobs(undefined, {
          type: ACTIVATE_JOB,
          job: {
            id: 1
          }
        })
      ).toMatchSnapshot();
    });

    it('should handle unknown action type', () => {
      expect(jobs(undefined, { type: 'unknown' })).toMatchSnapshot();
    });
  });
});
