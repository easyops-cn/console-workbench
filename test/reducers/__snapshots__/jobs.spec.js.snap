// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers jobs should handle ACTIVATE_JOB 1`] = `
Object {
  "active": 1,
  "entities": Object {},
  "ids": Array [],
  "outputs": Object {},
}
`;

exports[`reducers jobs should handle ACTIVATE_JOB again 1`] = `
Object {
  "active": null,
  "entities": Object {},
  "ids": Array [],
  "outputs": Object {},
}
`;

exports[`reducers jobs should handle ADD_JOB 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for ADD_JOB",
      "running": false,
      "starting": false,
      "stopping": false,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": undefined,
  },
}
`;

exports[`reducers jobs should handle CLEAR_JOB_OUTPUT 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for CLEAR_JOB_OUTPUT",
      "running": true,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": undefined,
  },
}
`;

exports[`reducers jobs should handle JOB_CLOSE 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for JOB_CLOSE",
      "running": false,
      "stopping": false,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [],
      "cursor": 0,
      "returned": "$ ls -l
drwxclosed with code: 15, signal: SIGTERM",
    },
  },
}
`;

exports[`reducers jobs should handle JOB_ERROR 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "error": [Error: error on purpose],
      "id": 1,
      "name": "test job for JOB_ERROR",
      "running": false,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [
        "drwx",
      ],
      "cursor": 4,
      "returned": "$ ls -l",
    },
  },
}
`;

exports[`reducers jobs should handle JOB_EXIT 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for JOB_EXIT",
      "running": false,
      "stopping": false,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [],
      "cursor": 0,
      "returned": "$ ls -l
drwxexited with code: 1, signal: SIGHUP",
    },
  },
}
`;

exports[`reducers jobs should handle JOB_OUTPUT 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for JOB_OUTPUT",
      "running": true,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [],
      "cursor": 0,
      "returned": "$ ls -l
drwx
100% done.",
    },
  },
}
`;

exports[`reducers jobs should handle REMOVE_JOB 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": undefined,
  },
  "ids": Array [],
  "outputs": Object {
    "1": undefined,
  },
}
`;

exports[`reducers jobs should handle REPLACE_JOBS 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for REPLACE_JOBS",
      "running": false,
      "starting": false,
      "stopping": false,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": undefined,
  },
}
`;

exports[`reducers jobs should handle START_JOB 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "error": undefined,
      "id": 1,
      "name": "test job for START_JOB",
      "starting": true,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [],
      "cursor": 0,
      "returned": "$ cd /tmp
$ ls -l",
    },
  },
}
`;

exports[`reducers jobs should handle STARTED_JOB 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for STARTED_JOB",
      "running": true,
      "starting": false,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [
        "drwx",
      ],
      "cursor": 4,
      "returned": "$ ls -l",
    },
  },
}
`;

exports[`reducers jobs should handle STOP_JOB 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -l",
      "cwd": "/tmp",
      "id": 1,
      "name": "test job for STOP_JOB",
      "running": true,
      "stopping": true,
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [],
      "cursor": 0,
      "returned": "$ ls -l
drwx^C",
    },
  },
}
`;

exports[`reducers jobs should handle UPDATE_JOB 1`] = `
Object {
  "active": null,
  "entities": Object {
    "1": Object {
      "cmd": "ls -la",
      "cwd": "/home",
      "id": 1,
      "name": "updated test job for UPDATE_JOB",
      "subPackageDir": "new-sub/pkg",
    },
  },
  "ids": Array [
    1,
  ],
  "outputs": Object {
    "1": Object {
      "buffer": Array [
        "drwx",
      ],
      "cursor": 4,
      "returned": "$ ls -l",
    },
  },
}
`;

exports[`reducers jobs should handle initial state 1`] = `
Object {
  "active": null,
  "entities": Object {},
  "ids": Array [],
  "outputs": Object {},
}
`;

exports[`reducers jobs should handle unknown action type 1`] = `
Object {
  "active": null,
  "entities": Object {},
  "ids": Array [],
  "outputs": Object {},
}
`;
